---
layout: post
title: "[.NET Concept]使用BeginXXX/EndXXX與SuspendLayout/ResumeLayout時，考慮加上Try/Finally"
date: 2010-07-18 12:48:29
comments: true
categories: [.NET Concept]
description: "[.NET Concept]使用BeginXXX/EndXXX與SuspendLayout/ResumeLayout時，考慮加上Try/Finally"
---
<p>相信大家都知道當在更新介面時，有的控制項會提供BeginUpdate/EndUpdate，甚至是BeginEdit/EndEdit、BeginInit/EndInit、SuspendLayout/ResumeLayout等暫停更新的方法，可用以加速介面的更新動作。像是ComboBox控制像就有這類方法：</p>  <div style="padding-bottom: 0px; margin: 0px; padding-left: 0px; padding-right: 0px; display: inline; float: none; padding-top: 0px" id="scid:812469c5-0cb0-4c63-8c15-c81123a09de7:94f18e86-0a3d-432e-aa2b-ae5e621f7131" class="wlWriterSmartContent"><pre name="code" class="vb">        ComboBox1.BeginUpdate()
        For i As Integer = 1 To 10000
            ComboBox1.Items.Add(i.ToString)
        Next
        ComboBox1.EndUpdate()</pre></div>

<p> </p>

<p>但若是未使用Try/Finally包住，當程式在呼叫EndUpdate前被莫名的原因中斷，像是例外發生。</p>

<div style="padding-bottom: 0px; margin: 0px; padding-left: 0px; padding-right: 0px; display: inline; float: none; padding-top: 0px" id="scid:812469c5-0cb0-4c63-8c15-c81123a09de7:8dab021a-2616-49a5-8c03-e81393914d44" class="wlWriterSmartContent"><pre name="code" class="vb">        ComboBox1.BeginUpdate()
        ...
        Throw New Exception
        ...
        ComboBox1.EndUpdate()</pre></div>

<p> </p>

<p>此時微軟預設的例外對話框會跳出，若使用者不願跳出系統因而按下繼續按鈕。</p>

<p><img style="border-right-width: 0px; border-top-width: 0px; border-bottom-width: 0px; border-left-width: 0px" border="0" alt="image" src="\images\posts\16611\image_thumb_1.png" width="444" height="157" /> </p>

<p> </p>

<p>或是您自行處理了一些事件把預設的例外框給取消掉。</p>

<div style="padding-bottom: 0px; margin: 0px; padding-left: 0px; padding-right: 0px; display: inline; float: none; padding-top: 0px" id="scid:812469c5-0cb0-4c63-8c15-c81123a09de7:ca59e1c0-fe2c-4355-be6c-a80f2a7f6192" class="wlWriterSmartContent"><pre name="code" class="vb">    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        AddHandler Application.ThreadException, AddressOf Application_ThreadException
        ...
        ComboBox1.BeginUpdate()
        ...
        Throw New Exception
        ...
        ComboBox1.EndUpdate()
    End Sub

    Private Sub Application_ThreadException(ByVal sender As Object, ByVal e As EventArgs)
        ...
    End Sub</pre></div>

<p> </p>

<p>整個介面在更新上就會出問題。</p>

<p><img style="border-right-width: 0px; border-top-width: 0px; border-bottom-width: 0px; border-left-width: 0px" border="0" alt="image" src="\images\posts\16611\image_thumb.png" width="203" height="88" /> </p>

<p> </p>

<p>要避免這樣的問題，必須養成適時為這些程式加上Try/Finally的好習慣，如此當中斷發生時就能確保介面仍能正常更新。</p>

<p />

<div style="padding-bottom: 0px; margin: 0px; padding-left: 0px; padding-right: 0px; display: inline; float: none; padding-top: 0px" id="scid:812469c5-0cb0-4c63-8c15-c81123a09de7:bc4a2da5-6a5e-4e5a-b2f0-547678918126" class="wlWriterSmartContent"><pre name="code" class="vb">        Try
            ComboBox1.BeginUpdate()
            ...
            Throw New Exception
            ...
        Finally
            ComboBox1.EndUpdate()
        End Try</pre></div>

<p />

<p><img style="border-right-width: 0px; border-top-width: 0px; border-bottom-width: 0px; border-left-width: 0px" border="0" alt="image" src="\images\posts\16611\image_thumb_2.png" width="203" height="88" /></p>